File: 02p1.py
reports = []
with open('inputs/02.csv') as f:
    f.readline() # skip the first line
    for line in f:
        reports.append([int(x) for x in line.split()])

safe = len(reports)

for report in reports:
    differences = []
    for i in range(0, len(report) - 1):
        differences.append(report[i] - report[i + 1])
    temp = True
    for difference in differences:
        if differences[0] < 0:
            if difference >= 0:
                print("Positive or zero distance in an increasing report: {}".format(differences))
                temp = False
                break
        elif differences[0] > 0:
            if difference <= 0:
                print("Negative or zero distance in a decreasing report: {}".format(differences))
                temp = False
                break
        else:
            print("Zero distance on first step: {}".format(differences))
            temp = False
            break
        if difference > 3 or difference < -3:
            print("Step too big: {}".format(differences))
            temp = False
            break
    if not temp:
        safe -= 1
    else:
        print("Safe: {}".format(differences))

print("Result: {}".format(safe))

File: 02p1.js
[
    {
      '$project': {
        '_id': 0, 
        'report': {
          '$map': {
            'input': {
              '$split': [
                '$reports', ' '
              ]
            }, 
            'in': {
              '$toInt': '$$this'
            }
          }
        }
      }
    }, {
      '$set': {
        'direction': {
          '$cond': {
            'if': {
              '$lt': [
                {
                  '$first': '$report'
                }, {
                  '$last': '$report'
                }
              ]
            }, 
            'then': 'increasing', 
            'else': 'decreasing'
          }
        }
      }
    }, {
      '$project': {
        'safe': {
          '$reduce': {
            'input': {
              '$range': [
                0, {
                  '$subtract': [
                    {
                      '$size': '$report'
                    }, 1
                  ]
                }
              ]
            }, 
            'initialValue': true, 
            'in': {
              '$let': {
                'vars': {
                  'distance': {
                    '$subtract': [
                      {
                        '$arrayElemAt': [
                          '$report', '$$this'
                        ]
                      }, {
                        '$arrayElemAt': [
                          '$report', {
                            '$add': [
                              '$$this', 1
                            ]
                          }
                        ]
                      }
                    ]
                  }
                }, 
                'in': {
                  '$cond': {
                    'if': {
                      '$not': '$$value'
                    }, 
                    'then': '$$value', 
                    'else': {
                      '$cond': {
                        'if': {
                          '$eq': [
                            '$direction', 'decreasing'
                          ]
                        }, 
                        'then': {
                          '$and': [
                            {
                              '$gt': [
                                '$$distance', 0
                              ]
                            }, {
                              '$lt': [
                                '$$distance', 4
                              ]
                            }
                          ]
                        }, 
                        'else': {
                          '$and': [
                            {
                              '$lt': [
                                '$$distance', 0
                              ]
                            }, {
                              '$gt': [
                                '$$distance', -4
                              ]
                            }
                          ]
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }, {
      '$match': {
        'safe': true
      }
    }, {
      '$count': 'res'
    }
  ]